Tasks/Maven/maven2.js

```diff
@@ -48,40 +51,149 @@ var publishJUnitResults = tl.getInput('publishJUnitResults');
-      tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');  
-      return 0;
+    console.log("SonarQube analysis is enabled");
+    var sqArguments;
+    var sqEndpoint = getEndpointDetails("sqConnectedServiceName");
+    var sqDbDetailsRequired = tl.getInput('sqDbDetailsRequired', true);
+
+    if (sqDbDetailsRequired == 'true') {
+        var sqDbUrl = tl.getInput('sqDbUrl', false);
+        var sqDbUsername = tl.getInput('sqDbUsername', false);
+        var sqDbPassword = tl.getInput('sqDbPassword', false);
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password, sqDbUrl, sqDbUsername, sqDbPassword);
+    } else {
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password);
     }
 
-    var tp = new tl.TestPublisher("JUnit");
-    tp.publish(matchingTestResultsFiles, false, "", "");
-  } 
+    sqArguments = mavenOptions + " " + sqArguments;
+    tl.debug("Running Maven with goal sonar:sonar and options " + sqArguments);
+
+    var mvnsq = new tl.ToolRunner(mvntool);
+    mvnsq.arg('-f');
+    mvnsq.arg(mavenPOMFile);
+    mvnsq.arg(sqArguments);
+    mvnsq.arg("sonar:sonar");
+
+    return mvnsq;
 }
 
+function getEndpointDetails(inputFieldName) {
+    var errorMessage = "Could not decode the generic endpoint. Please ensure you are running the latest agent (min version 0.3.0)"
+    if (!tl.getEndpointUrl) {
+        throw new Error(errorMessage);
+    }
+
+    var genericEndpoint = tl.getInput(inputFieldName);
+    if (!genericEndpoint) {
+        throw new Error(errorMessage);
+    }
+
+    hostUrl = tl.getEndpointUrl(genericEndpoint, false);
+    var auth = tl.getEndpointAuthorization(genericEndpoint, false);
+
+    if (auth.scheme != "UsernamePassword") {
+        throw new Error("The authorization scheme " + auth.scheme + " is not supported for a SonarQube endpoint. Please use a username and a password.");
+    }
+
+    hostUsername = auth.parameters.Username;
+    hostPassword = auth.parameters.Password;
+
+    return {
+        "Url": hostUrl,
+        "Username": hostUsername,
+        "Password": hostPassword
+    };
+}
+
+function createSonarQubeArgs(sqHostUrl, sqHostUsername, sqHostPassword, sqDbUrl, sqDbUsername, sqDbPassword) {
+    var sqArgs = "";
+    if (sqHostUrl) {
+        sqArgs += '-Dsonar.host.url="' + sqHostUrl + '" '
+    }
+    if (sqHostUsername) {
+        sqArgs += '-Dsonar.login="' + sqHostUsername + '" '
+    }
+    if (sqHostPassword) {
+        sqArgs += '-Dsonar.password="' + sqHostPassword + '" '
+    }
+    if (sqDbUrl) {
+        sqArgs += '-Dsonar.jdbc.url="' + sqDbUrl + '" '
+    }
+    if (sqDbUsername) {
+        sqArgs += '-Dsonar.jdbc.username="' + sqDbUsername + '" '
+    }
+    if (sqDbPassword) {
+        sqArgs += '-Dsonar.jdbc.password="' + sqDbPassword + '" '
+    }
+
+    return sqArgs;
+}
+
+// Maven task orchestration
```

Comment:
We cannot differentiate between a build and a test failure - the maven task simply fails. The SQ analysis picks up any the information it can - if the build is broken it will only look at the source code. If some tests fail, SQ will pick the source code, the binaries and the test results. Note that I confirmed that this is the "right thing to do" with Dinesh, i.e. if the build fails we should still do a SQ analysis.